### PR TITLE
Prevents viewer from using the video seekbar to skip questions when skip disabled

### LINKED ISSUE #628: Viewers can use the video seekbar to skip questions when skip disabled

### PR DESCRIPTION
**Pull request by dalmia**
Fixes #628 

## Summary
- maintaining a variable for the last answered interaction index
- when the video is seeked, if the skip is disabled and there are interactions pending and the timestamp that is being seeked to is greater than the timestamp of the first unanswered interaction index, the viewer is shown that interaction
- when the page is reloaded, if the user was at a timestamp > timestamp of the first unanswered interaction index, we show that interaction.
- if a user clicks on play when on an unanswered question, they are not allowed to move ahead if skip is disabled
 
## Test Plan

- [ ] ~Test Responsiveness~
   - [ ] Laptop (1200px)
   - [ ] Tablet (760px)
   - [ ] Phone (320px)
- [x] Cross-Browser Testing
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Safari (use BrowserStack)
- [x] Local Language Support
- [ ] Hygiene and Housekeeping
   - [ ] Self-review
   - [ ] Comments have been added appropriately
   - [ ] ~Check for bundle size [here](https://bundlephobia.com/) if adding a package~
   - [x] Added relevant details like Labels/Projects/Milestones etc.
   - [ ] ~If adding or removing any environment variable:~
       - [ ] update `docs/ENV.md`
       - [ ] update Github Workflow files
       - [ ] update DockerFile
       - [ ] update the secrets for staging and production
- [ ] Testing
   - [ ] Wrote tests
   - [ ] Tested locally
   - [ ] Tested on staging
   - [ ] Tested on an actual physical phone
   - [ ] Tested on production
- [ ] ~Lighthouse Checks~
   - [ ] Images have `alt` attributes
   - [ ] Any `` tags have `width` and `height` specified
   - [ ] Any `target="_blank"` links have `rel="noopener"`
   - [ ] Only SVGs are used as images. If PNGs are used, their size has been optimised.
   - [ ] Any SVGs without text have their `aria-label` attributes set

### CODE REVIEW

**Comment by deepansh96** [comment]
- There are two places where the tooltip comes on top of the active div in touch/mobile. One is when I click the `Preview` button. Once is when I click `Publish` button. Can you fix this in this PR itself? Should be a small change

- I had this weird bug on mobile phone where in preview mode and in player as well, the questions were not getting updated to green color after answering. I guess this is not because of this PR, but due to this, I was being redirected to the same question again and again because the code thought that no question has been answered.

https://user-images.githubusercontent.com/20169829/155518919-7e02378c-3ab6-4dc9-8000-0b44eae6321d.mp4


- What should we do for cases where, let's say someone has answered a plio where skip was enabled. They answered some questions and skipped some questions. Now, the creator changed the setting to "skip enabled" as false. Now when the user returns to the plio, the code will drop the user at the last unanswered question, even if there are some other unanswered questions in between. The code is written with the assumption that all questions will be answered in order. So I'm not sure what should the code ideally do in this case.


- Also, a separate bug, not for this PR. In subjective question textbox, if I keep typing something without any space, the submit button stays hidden idk why. Only if I press spacebar or close the mobile keyboard, the submit button gets active

https://user-images.githubusercontent.com/20169829/155519290-f9ff0e46-0996-4969-b94a-143c84a8ba93.mp4

- bug, not always replicable. But replicable on mobile phones. Sometimes, even if the coloring of the markers is correct, it doesn't allow me to go back into the green questions waala area whereas it should allow me. And sometimes it starts working fine.

https://user-images.githubusercontent.com/20169829/155519867-b279d80d-5a64-43aa-b08d-a7451b051621.mp4

File: src/pages/Embeds/Plio.vue [hunk 1]
```diff
@@ -659,7 +686,47 @@ export default {
...
+        timeToInspect
+      );
+      if (firstUnansweredInteraction != null) {
+        this.setPlayerTime(firstUnansweredInteraction.time - POP_UP_CHECKING_FREQUENCY);
+        setTimeout(() => {
+          this.toast.error(this.$t("toast.player.cannot_skip_interaction"), {
+            id: "cannotSkipInteraction",
+            position: "bottom-center",
+          });
+        }, 500);
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:719]: The settimeout here means that after the player time is set above, 500ms after it, the toast would pop up?
This works fine mostly but on a slow phone (shivani's office phone) what happens is that the toast pops up before.

When you do `this.setPlayerTime(`, plyr tries to buffer the video from that point. Until some amount has been buffered, the question modal that should pop up, doesn't pops up. 

Wouldn't it be a better solution to show this toast when an item has popped up instead of relying on 500ms?

File: src/pages/Embeds/Plio.vue [hunk 2]
```diff
@@ -836,8 +907,26 @@ export default {
           if (this.isItemMCQ(itemIndex)) {
             itemResponse.answer = parseInt(itemResponse.answer);
           }
+          if (
+            !this.isItemResponseEmpty(itemIndex, itemResponse.answer) &&
+            !this.isSkipEnabled
+          ) {
+            this.lastAnsweredInteractionIndex = itemIndex;
+          }
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:915]: This piece will cause some confusion in one of the edge cases that I've mentioned in the list of comments I've left on the PR. 
This assumes that no unanswered question can exist before an answered question?
Because in the case where there are 4 unanswered questions and then 1 answered question, it will make the last question as `lastAnsweredItemIndex` and the code will never force the user to answer the other questions.

I don't know what the solution should be.
Either we disallow people to change this setting after they publish as it can cause conflict in the responses.
Or we warn them that some older users will not be forced to answer.
Or we update this code to actually force the user to answer all the left out questions in this specific case

File: src/pages/Embeds/Plio.vue [hunk 3]
```diff
@@ -836,8 +907,26 @@ export default {
...
+            this.currentTimestamp
+          );
+          // check if any item before the current timestamp is unanswered
+          // if there is, then update the timestamp to just before the
+          // unanswered item
+          if (firstUnansweredInteraction != null) {
+            this.currentTimestamp =
+              firstUnansweredInteraction.time - POP_UP_CHECKING_FREQUENCY;
+          }
+        }
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:929]: This logic is what is needed to be checked everytime the video is seeked
Currently you're doing it when creating session, but it should also be run whenever it is seeked. This will solve the bug that I experienced on mobile phones (added video clip in the comments on the PR)

File: src/pages/Embeds/Plio.vue [hunk 4]
```diff
@@ -1043,23 +1137,26 @@ export default {
...
 
-      this.checkForItemPopup(timestamp);
+      const itemIndex = this.checkForItemPopup(timestamp);
+      if (itemIndex != null) {
+        this.currentItemIndex = itemIndex;
+        this.markItemSelected();
+        this.createEvent("item_opened", { itemIndex: this.currentItemIndex });
+      } else {
+        this.closeItemModal();
+      }
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:1147]: doubt - why was this added?

File: src/pages/Embeds/Plio.vue [hunk 5]
```diff
@@ -1043,23 +1137,26 @@ export default {
...
-      this.currentItemIndex = ItemFunctionalService.checkItemPopup(
+      return ItemFunctionalService.checkItemPopup(
         timestamp,
         this.itemTimestamps,
         POP_UP_PRECISION_TIME
       );
-      if (this.currentItemIndex != null) {
-        this.markItemSelected();
-        this.createEvent("item_opened", { itemIndex: this.currentItemIndex });
-      }
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:1062]: doubt - why was this removed?

**Review by deepansh96 — CHANGES_REQUESTED** [review]

File: src/pages/Embeds/Plio.vue [hunk 6]
```diff
@@ -202,6 +205,7 @@ export default {
       isAspectRatioChecked: false, // whether the check for aspect ratio has been done
       watchingEventDBId: null, // the DB id of the latest 'watching' event for a given session
       plioSettings: null, // stores this plio's settings
+      lastAnsweredInteractionIndex: -1, // index of the interaction that was last answered
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:208]: haan sorry makes sense! 👍

File: src/pages/Embeds/Plio.vue [hunk 7]
```diff
@@ -511,7 +532,12 @@ export default {
   },
   methods: {
     ...mapActions("auth", ["setAccessToken", "setActiveWorkspace"]),
-
+    isItemResponseEmpty(itemIndex, answer) {
+      if (answer == null) return true;
+      if (this.itemDetails[itemIndex].type == "mcq") return isNaN(answer);
+      if (this.itemDetails[itemIndex].type == "checkbox") return answer.length == 0;
+      return false;
+    },
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:540]: uska is already covered in `answer == null` - warna it is non-empty - this is how it is implemented everywhere else as well jahan bhi we are checking if answer has been submitted or not.

File: src/pages/Embeds/Plio.vue [hunk 8]
```diff
@@ -562,6 +588,7 @@ export default {
      * Show the scorecard on top of the player
      */
     popupScorecard() {
+      if (this.checkMovingToTimestampAllowed(this.player.duration) != null) return;
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:591]: haan naam I will change (good suggestion 👍)

boolean nahi hain kyunki it returns an object corresponding to the last answered item if there is any + if skip is disabled!

[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:591]: without this, if the user drags the seekbar to the video ka end, toh before our check can run, the scorecard is popped up beyond for the scorecard pop-up there is no check to be run - plyr automatically triggers it as we reach the video end.

File: src/pages/Embeds/Plio.vue [hunk 9]
```diff
@@ -659,7 +686,47 @@ export default {
         if (maximizeButton != undefined) this.mountOnFullscreenPlyr(maximizeButton);
       });
     },
+    checkMovingToTimestampAllowed(timestamp) {
+      if (!this.isSkipEnabled && this.lastAnsweredInteractionIndex < this.numItems - 1) {
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:690]: since ek hi jagah use ho raha - is it okay if I add a comment here instead of adding another variable to the global scope?

File: src/pages/Embeds/Plio.vue (hunk 4 above)
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:1147]: just for cleanup - I was earlier trying to use `checkForItemPopup` and it was doing more things than just checking for item popup - so, was hard to reuse - abhi although I didn't end up using `checkForItemPopup` as I was originally trying to, this cleanup still made sense, no?

File: src/pages/Embeds/Plio.vue (hunk 5 above)
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:1062]: explained above.

File: src/pages/Embeds/Plio.vue [hunk 10]
```diff
@@ -659,7 +686,47 @@ export default {
...
+        this.player.currentTime < this.currentItem.time &&
+        this.player.currentTime >= this.currentItem.time - POP_UP_CHECKING_FREQUENCY
+      )
+        timeToInspect += POP_UP_CHECKING_FREQUENCY;
+
+      const firstUnansweredInteraction = this.checkMovingToTimestampAllowed(
+        timeToInspect
+      );
+      if (firstUnansweredInteraction != null) {
+        this.setPlayerTime(firstUnansweredInteraction.time - POP_UP_CHECKING_FREQUENCY);
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:713]: because our check for whether item should be shown or not checks for a window of POP_UP_CHECKING_FREQUENCY **before** the actual timestamp. Probably doing `-0.1` or something should also work but would be more unintuitive. should I add a comment here?

**Comment by dalmia** [comment]
> There are two places where the tooltip comes on top of the active div in touch/mobile. One is when I click the Preview button. Once is when I click Publish button. Can you fix this in this PR itself? Should be a small change

> Also, a separate bug, not for this PR. In subjective question textbox, if I keep typing something without any space, the submit button stays hidden idk why. Only if I press spacebar or close the mobile keyboard, the submit button gets active

made a separate issue for these two - any of us can solve them in a separate PR - don't want to further clutter this one.

**Comment by dalmia** [comment]
> I had this weird on mobile phone where in preview mode and in player as well, the questions were not getting updated to green color after answering. I guess this is not because of this PR, but due to this, I was being redirected to the same question again and again because the code thought that no question has been answered.

preview mode mein we are not recording the question ka answers kabhi. toh whenever you close preview mode, the answers are reset. no session is created for preview mode ka responses na? player ke saath toh nahi hona chaiye - can you actually check if that is happening again?

**Comment by dalmia** [comment]
> What should we do for cases where, let's say someone has answered a plio where skip was enabled. They answered some questions and skipped some questions. Now, the creator changed the setting to "skip enabled" as false. Now when the user returns to the plio, the code will drop the user at the last unanswered question, even if there are some other unanswered questions in between. The code is written with the assumption that all questions will be answered in order. So I'm not sure what should the code ideally do in this case.

haan this is a valid concern. instead of restricting them from updating after published, we can simply set the last unanswered index to the last index before any unanswered questions begin - that makes sense? the reason is that I want to really avoid restricting people for edge case scenarios if they can be handled by us in a decent manner.

File: src/pages/Embeds/Plio.vue (hunk 3 above)
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:929]: videoSeeked wala function mein we are calling `moveToFirstUnansweredItemTimestampOrPass` right? wahan pe ho raha same check!

File: src/pages/Embeds/Plio.vue (hunk 2 above)
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:915]: haan made a suggestion towards the end of the comments. let me know what you think!

File: src/pages/Embeds/Plio.vue (hunk 1 above)
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:719]: haan the trick is to decide if the toast is to be shown in a different function than where this check was done - the function that pops up the item - thik hain - will have a look!

File: src/pages/Embeds/Plio.vue [hunk 11]
```diff
@@ -659,7 +686,47 @@ export default {
...
     },
+    checkMovingToTimestampAllowed(timestamp) {
+      if (!this.isSkipEnabled && this.lastAnsweredInteractionIndex < this.numItems - 1) {
+        const firstUnansweredInteraction = this.items[
+          this.lastAnsweredInteractionIndex + 1
+        ];
+        if (timestamp >= firstUnansweredInteraction.time)
+          return firstUnansweredInteraction;
+      }
+    },
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:697]: haan nice suggestion 👍

File: src/pages/Embeds/Plio.vue [hunk 12]
```diff
@@ -202,6 +209,8 @@ export default {
       isAspectRatioChecked: false, // whether the check for aspect ratio has been done
       watchingEventDBId: null, // the DB id of the latest 'watching' event for a given session
       plioSettings: null, // stores this plio's settings
+      lastAnsweredItemIndex: -1, // index of the interaction that was last answered
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:212]:
```suggestion
      lastAnsweredItemIndex: -1, // index of the item that was last answered
```

File: src/pages/Embeds/Plio.vue [hunk 13]
```diff
@@ -656,12 +700,53 @@ export default {
...
         let maximizeButton = document.getElementById("plioMaximizeButton");
-        if (maximizeButton != undefined) this.mountOnFullscreenPlyr(maximizeButton);
+        if (maximizeButton != undefined)
+          this.mountOnFullscreenPlyr(maximizeButton);
       });
     },
+    isMovingToTimestampAllowed(timestamp) {
+      if (this.firstUnansweredItem == null) return true;
+      // check whether skip is disabled and all items have been answered
+      return timestamp < this.firstUnansweredItem.time;
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:710]: here the comment says that `check whether skip is disabled...` but that check is not happening in the code?

File: src/pages/Embeds/Plio.vue [hunk 14]
```diff
@@ -656,12 +700,53 @@ export default {
...
+    },
+    /**
+     * whether the user should be moved to the first unanswered item
+     * or allowed to pass through to whatever timestamp they are trying to
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:714]:
```suggestion
     * or is allowed to pass through to whatever timestamp they are trying to
```

File: src/pages/Embeds/Plio.vue [hunk 15]
```diff
@@ -656,12 +700,53 @@ export default {
       // insert the button inside the plyr instance so that it shows up in fullscreen mode
       this.$nextTick(() => {
         let maximizeButton = document.getElementById("plioMaximizeButton");
-        if (maximizeButton != undefined) this.mountOnFullscreenPlyr(maximizeButton);
+        if (maximizeButton != undefined)
+          this.mountOnFullscreenPlyr(maximizeButton);
       });
     },
+    isMovingToTimestampAllowed(timestamp) {
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:707]: add @param for timestamp?

File: src/pages/Embeds/Plio.vue [hunk 16]
```diff
@@ -511,7 +544,13 @@ export default {
   },
   methods: {
     ...mapActions("auth", ["setAccessToken", "setActiveWorkspace"]),
-
+    isItemResponseEmpty(itemIndex, answer) {
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:547]: add @param for `itemIndex` and `answer`

File: src/pages/Embeds/Plio.vue [hunk 17]
```diff
@@ -656,12 +700,53 @@ export default {
...
+      if (this.isSkipEnabled) return false;
+      let timeToInspect = this.player.currentTime;
+
+      /**
+       * this is required to handle the case that the user toggles
+       * the show video/question button while an item is active and tries
+       * to play the video since the user will be in the time range of item
+       * timestamp and item timestamp - POP_UP_CHECKING_FREQUENCY, the code
+       * after this snippet will function incorrectly
+       */
```
Comment (deepansh96) [inline src/pages/Embeds/Plio.vue:727]:
```suggestion
      /**
       * this is required to handle the case when the user toggles
       * the show video/question button while an item is active and tries
       * to play the video. 
       * since the user will be in the time range of item timestamp and 
       * item timestamp - POP_UP_CHECKING_FREQUENCY, the code
       * after this snippet will function incorrectly
       */
```

**Review by deepansh96 — CHANGES_REQUESTED** [review]

**Review by deepansh96 — APPROVED** [review]
